### PR TITLE
[iOS][GPU] Add the iOS GPU workflow recipe

### PR DESCRIPTION
**Pull request by xta0**

### CODE REVIEW

File: prototype_source/ios_gpu_workflow.rst [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+You may notice that the results are slighly different from the `results <https://pytorch.org/mobile/ios/#install-libtorch-via-cocoapods>`_ we got from the CPU model as shown in the iOS tutorial. This because by default Metal uses fp16 rather than fp32 to compute. The precision loss from the result is expected.
+
+
+Conclusion
+----------
+
+In this tutorial, we demonstrated how to convert the existing torchscript model to a GPU compatible model for iOS. We walked through the HelloWorld example to show how to use the C++ APIs. Please be aware of that GPU feature is still under development, new operators will continue be added. APIs are subject to change in the future versions.
+
+Thanks for reading! As always, we welcome any feedback, so please create an issue `here <https://github.com/pytorch/pytorch/issues>`_ if you have any.
+
```
Comment (jlin27) [inline prototype_source/ios_gpu_workflow.rst:99]: Looks good! Only comment would be are there any related APIs docs or models (e.g. [Mobile Net v2](https://pytorch.org/hub/pytorch_vision_mobilenet_v2/)) that can be linked either in a "Learn More" section at the end or in the introduction?

**Review by jlin27 — CHANGES_REQUESTED** [review]

File: prototype_source/ios_gpu_workflow.rst [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+==================================
+
+**Author**: `Tao Xu <https://github.com/xta0>`_
+
+Introduction
+------------
+
+This tutorial introduces the steps to run your models on iOS GPU. We'll be using the mobilenetv2 model as an example. Since the mobile GPU features are currently in the prototype stage, you'll need to build a custom pytorch binary from source. For the time being, only a limited number of operators are supported, and certain client side APIs are subject to change in the future versions.
+
+Mobile Preparation
```
Comment (jeffxtang) [inline prototype_source/ios_gpu_workflow.rst:11]: Mobile -> Model

File: prototype_source/ios_gpu_workflow.rst [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+-------------------
+
+Since GPUs consume weights in a different order, the first step we have to do is to convert our torchscript mobile to a GPU compatible model. This step is also known as "prepacking". To do that, we'll build a custom pytorch binary from source that includes the Metal backend. Go ahead checkout the pytorch source code from github and run the command below
```
Comment (jeffxtang) [inline prototype_source/ios_gpu_workflow.rst:14]: torchscript mobile -> TorchScript model

Comment (jeffxtang) [reply prototype_source/ios_gpu_workflow.rst:14]: or keep torchscript lower case to be consistent.

File: prototype_source/ios_gpu_workflow.rst [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    ['aten::adaptive_avg_pool2d', 
+    'aten::add.Tensor', 
+    'aten::addmm', 
+    'aten::reshape', 
+    'aten::size.int', 
+    'metal::copy_to_host', 
+    'metal_prepack::conv2d_run']
+
+Those are all the ops we need to run the mobilenetv2 model on iOS GPU. Cool! Now you have the ``mobilenetv2_metal.pt`` saved on your disk, let's move on to the iOS part.
```
Comment (jeffxtang) [inline prototype_source/ios_gpu_workflow.rst:54]: Now -> Now that

File: prototype_source/ios_gpu_workflow.rst [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+-------------------
+
+Since GPUs consume weights in a different order, the first step we have to do is to convert our torchscript mobile to a GPU compatible model. This step is also known as "prepacking". To do that, we'll build a custom pytorch binary from source that includes the Metal backend. Go ahead checkout the pytorch source code from github and run the command below
+
+.. code:: shell
+
+    cd PYTORCH_ROOT
+    USE_PYTORCH_METAL=ON python setup.py install --cmake
+
+The command above will build a custom pytorch binary from master. The ``install`` argument simply tells ``setup.py`` to override the existing PyTorch on your desktop. Once the build finished, I would recommend opening another terminal and checking the PyTorch version to see if the installation successful. As the time of writing of this recipe, the version is ``1.8.0a0+41237a4``. The number might be different depending on when you check out the code from master, but it should be greater than 1.7.0.
```
Comment (jeffxtang) [inline prototype_source/ios_gpu_workflow.rst]: Works - confirmed 👍

File: prototype_source/ios_gpu_workflow.rst [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+
+The next step is going to be converting the mobilenetv2 torchscript model to a Metal compatible model. We'll be leveraging the ``optimize_for_mobile`` API from ``torch.utils`` as shown below
+
+.. code:: python
+
+    import torch
+    from torch.utils.mobile_optimizer import optimize_for_mobile
+
+    scripted_model = torch.jit.load('./mobilenetv2.pt')
```
Comment (jeffxtang) [inline prototype_source/ios_gpu_workflow.rst:36]: replace this line with:
```
import torchvision

model_quantized = torchvision.models.quantization.mobilenet_v2(pretrained=True, quantize=True)
scripted_model = torch.jit.script(model_quantized)
```

or the non-quantized version:
```
import torchvision

model = torchvision.models.mobilenet_v2(pretrained=True)
scripted_model = torch.jit.script(model)
```

and the user can then copy and paste the while code snippet and run without getting an error "mobilenetv2.pt does not exist" and etc..

File: prototype_source/ios_gpu_workflow.rst [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    optimized_model = optimize_for_mobile(scripted_model, backend='metal')
+    torch.jit.export_opnames(optimized_model)
```
Comment (jeffxtang) [inline prototype_source/ios_gpu_workflow.rst:38]: should have a `print` around: `print(torch.jit.export_opnames(optimized_model))`

File: prototype_source/ios_gpu_workflow.rst [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    torch.jit.save(optimized_model, './mobilenetv2_metal.pt')
+
+Note that the ``torch.jit.export_opnames(optimized_model)`` is going to dump all the optimized operators from the ``optimized_mobile``. If everything works well, you should be able to see the following ops printed out
+
+
+.. code:: shell
```
Comment (jeffxtang) [inline prototype_source/ios_gpu_workflow.rst:44]: My output for using the non-quantized model above is: `
['aten::__getitem__.t', 'aten::adaptive_avg_pool2d', 'aten::add.Tensor', 'aten::add_.Tensor', 'aten::addmm', 'aten::dim', 'aten::eq.int', 'aten::gt.int', 'aten::len.t', 'aten::matmul', 'aten::reshape', 'aten::size', 'metal::copy_to_host', 'metal_prepack::conv2d_run', 'prim::RaiseException']`

File: prototype_source/ios_gpu_workflow.rst [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    'metal::copy_to_host', 
+    'metal_prepack::conv2d_run']
+
+Those are all the ops we need to run the mobilenetv2 model on iOS GPU. Cool! Now you have the ``mobilenetv2_metal.pt`` saved on your disk, let's move on to the iOS part.
+
+
+Use C++ APIs
+---------------------
+
+In this section, we'll be using the `HelloWorld example <https://github.com/pytorch/ios-demo-app>`_ to demonstrate how to use the C++ APIs to run our model. The first thing we need to do is to build a custom iOS LibTorch from Source. Make sure you have deleted the **build** folder from the previous step in PyTorch root directory. Then run the command below
```
Comment (jeffxtang) [inline prototype_source/ios_gpu_workflow.rst]: Make sure you have deleted ->
Run `rm -rf build` to delete

File: prototype_source/ios_gpu_workflow.rst [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+.. code:: shell
+    
+    IOS_ARCH=arm64 ./scripts/build_ios.sh -DUSE_PYTORCH_METAL=ON
+
+Note `IOS_ARCH` tells the script to build a arm64 version of Libtorch. This is because the PyTorch Metal is only available for the iOS devices that support the Apple A9 chip or above. Once the build finished, follow the `Build PyTorch iOS libraries from source <https://pytorch.org/mobile/ios/#build-pytorch-ios-libraries-from-source>`_ section from the iOS tutorial to setup the XCode settings properly. Also, don't forget to copy the `./mobilenetv2_metal.pt` to your XCode project.
```
Comment (jeffxtang) [inline prototype_source/ios_gpu_workflow.rst]: "follow the `Build PyTorch iOS libraries from source <https://pytorch.org/mobile/ios/#build-pytorch-ios-libraries-from-source>`_ section" is a little confusing as it starts with BUILD PYTORCH IOS LIBRARIES FROM SOURCE, Build LibTorch for iOS Simulators, and Build LibTorch for arm64 Devices. Should change that to:
follow the `Xcode Setup <https://pytorch.org/mobile/ios/#xcode-setup`_ section

File: prototype_source/ios_gpu_workflow.rst [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+
+Use C++ APIs
+---------------------
+
+In this section, we'll be using the `HelloWorld example <https://github.com/pytorch/ios-demo-app>`_ to demonstrate how to use the C++ APIs to run our model. The first thing we need to do is to build a custom iOS LibTorch from Source. Make sure you have deleted the **build** folder from the previous step in PyTorch root directory. Then run the command below
+
+.. code:: shell
+    
+    IOS_ARCH=arm64 ./scripts/build_ios.sh -DUSE_PYTORCH_METAL=ON
```
Comment (jeffxtang) [inline prototype_source/ios_gpu_workflow.rst:64]: This works - i can see the build_ios/install

File: prototype_source/ios_gpu_workflow.rst (hunk 10 above)
Comment (jeffxtang) [reply prototype_source/ios_gpu_workflow.rst]: I'm having problem building the HelloWorld example using the include and lib files built above. That Xcode Setup section is not very detailed.. I set the Header Search Path in Xcode but still get the LibTorch/LibTorch.h file not found error..

File: prototype_source/ios_gpu_workflow.rst (hunk 8 above)
Comment (xta0) [reply prototype_source/ios_gpu_workflow.rst:44]: I'm afraid that the quantized model won't work. Have to use the float model.
